otherlibs/stdune-unstable/user_error.mli

```diff
@@ -1,15 +1,37 @@
 (** Error meant for humans *)
 
+module Annot : sig
+  type t
+
+  module type S = sig
+    type payload
+
+    val make : payload -> t
+
+    val check : t -> (payload -> 'a) -> (unit -> 'a) -> 'a
```

Comment:
We do something very similar in `Dune_util.Persistent`, but IMO using a clearer API. The idea is to use a  "private extensible type":

```
module Annot : sig
  type t = private ..

  module type S = sig
    type payload
    type t += T of payload

    val make : payload -> t
  end
```

With this API, the user can simply pattern match `t` values but still cannot add new constructors to `t` without going through `Make`.